src/pg_copy.c

```diff
@@ -0,0 +1,626 @@
+
+	/*
+	 * Similarly, line_buf holds the whole input line being processed. The
+	 * input cycle is first to read the whole line into line_buf, convert it
+	 * to server encoding there, and then extract the individual attribute
+	 * fields into attribute_buf.  line_buf is preserved unmodified so that we
+	 * can display it in error messages if appropriate.
+	 */
+	StringInfoData line_buf;
+	bool		line_buf_converted;		/* converted to server encoding? */
+	bool		line_buf_valid; /* contains the row being processed? */
+
+	/*
+	 * Finally, raw_buf holds raw data read from the data source (file or
+	 * client connection).	CopyReadLine parses this data sufficiently to
+	 * locate line boundaries, then transfers the data to line_buf and
+	 * converts it.	 Note: we guarantee that there is a \0 at
+	 * raw_buf[raw_buf_len].
+	 */
+#define RAW_BUF_SIZE 65536		/* we palloc RAW_BUF_SIZE+1 bytes */
+	char	   *raw_buf;
+	int			raw_buf_index;	/* next byte to process */
+	int			raw_buf_len;	/* total # of bytes stored */
+} CopyStateData;
+
+/* TODOL should be moved to copy.c */
+static StringInfo CopyGetLineBuf(CopyStateData *cs)
+{
+	return &cs->line_buf;
+}
+
+#define MAX_SHARDS 1001
+#define MAX_REPLICATION_FACTOR 8 /* should be less or equal than 32 */
+#define MAX_STMT_LEN 1024
+
+typedef struct {
+	int64 shardId;
+	int nReplicas;
+	int preparedMask;
+	PGconn* conn[MAX_REPLICATION_FACTOR];
+} CopyConnection;
+
+static uint32 shard_id_hash_fn(const void *key, Size keysize)
+{
+	return *(int32*)key ^ *((int32*)key+1);
+}
+
+/* 
+ * Construct hash table used for shardId->Connection mapping.
+ * We can not use connection cache from connection.c used by GteConnection because
+ * we need to establish multiple connections with each nodes: one connection per shard
+ */
+static HTAB *
+CreateShard2ConnectionHash()
+{
+	HASHCTL info;
+
+	memset(&info, 0, sizeof(info));
+	info.keysize = sizeof(int64);
+	info.entrysize = sizeof(CopyConnection);
+	info.hash = shard_id_hash_fn;
+
+	return hash_create("shard2conn", MAX_SHARDS, &info, HASH_ELEM | HASH_FUNCTION);
+}
+
+/*
+ * Execute statement with specified parameters and check its result
+ */
+static bool PgShardExecute(PGconn* conn, ExecStatusType expectedResult, char const* sql, ...)
+{
+	PGresult *result;
+	bool ret = true;
+	va_list args;
+	char buf[MAX_STMT_LEN];
+
+	va_start(args, sql);
+	vsnprintf(buf, sizeof(buf), sql, args);
+	result = PQexec(conn, buf);
+
+	if (PQresultStatus(result) != expectedResult)
+	{
+		ReportRemoteError(conn, result);
+		ret = false;
+	}
+	return ret;
+}
+
+/*
+ * Handle copy to/from distributed table
+ */
+bool PgShardCopy(CopyStmt *copyStatement, char const* query)
+{		 
+	RangeVar *relation = copyStatement->relation;
+	ListCell *shardIntervalCell;
+	ListCell *taskPlacementCell;
+	List *shardIntervalList;
+	int relationSuffixPos;
+	char *relationName;
+	char const *lowerQuery, *relationOcc;
+	PgCopyTransactionManager* tmgr = &PgCopyTmgrImpl[PgCopyTMGR];
+
+	if (relation != NULL)
```

Comment:
To avoid deep nesting, how about we return if relation == NULL? Or do this check prior to calling PgShardCopy? It seems the code for that was already there.